openedx/core/lib/block_cache/block_structure.py

```diff
@@ -0,0 +1,588 @@
+                transformer data is to be updated.
+
+            transformer (BlockStructureTransformer) - The transformer
+                whose data is to be updated.
+
+            key (string) - A dictionary key to the transformer's data.
+
+            value (any picklable type) - The value to associate with the
+                given key for the given transformer's data for the
+                requested block.
+        """
+        self._block_data_map[usage_key]._transformer_data[transformer.name()][key] = value
+
+    def remove_transformer_block_data(self, usage_key, transformer, key):
+        """
+        Deletes the given transformer's data dictionary value for the
+        given key for the block identified by the given usage_key.
+
+        Arguments:
+            usage_key (UsageKey) - Usage key of the block whose
+                transformer data is to be deleted.
+
+            transformer (BlockStructureTransformer) - The transformer
+                whose data entry is to be deleted.
+
+            key (string) - The key to the transformer's dictionary entry
+                that is to be deleted.
+        """
+        self._block_data_map[usage_key]._transformer_data.get(transformer.name(), {}).pop(key, None)
+
+    def remove_block(self, usage_key, keep_descendants):
+        """
+        Removes the block identified by the usage_key and all of its
+        related data from the block structure.  If descendants of the
+        removed block are to be kept, the structure's relations are
+        updated to reconnect the block's parents with its children.
+
+        Note: While the immediate relations of the block are updated
+        (removed), all descendants of the block will remain in the
+        structure unless the _prune_unreachable method is called.
+
+        Arguments:
+            usage_key (UsageKey) - Usage key of the block that is to be
+                removed.
+
+            keep_descendants (bool) - If True, the block structure's
+                relations (graph edges) are updated such that the
+                removed block's children become children of the
+                removed block's parents.
+        """
+        children = self._block_relations[usage_key].children
+        parents = self._block_relations[usage_key].parents
+
+        # Remove block from its children.
+        for child in children:
+            self._block_relations[child].parents.remove(usage_key)
+
+        # Remove block from its parents.
+        for parent in parents:
+            self._block_relations[parent].children.remove(usage_key)
+
+        # Remove block.
+        self._block_relations.pop(usage_key, None)
+        self._block_data_map.pop(usage_key, None)
+
+        # Recreate the graph connections if descendants are to be kept.
+        if keep_descendants:
+            [self._add_relation(parent, child) for child in children for parent in parents]  # pylint: disable=expression-not-assigned
```

Comment:
Why not just use normal loops instead of:
    
1. Creating a list that then just gets discarded.
2. Uses a double list comprehension (generally a code smell, though not so bad here, since there's no dependency between the two variables)
3. Forces you to disable a pylint warning.